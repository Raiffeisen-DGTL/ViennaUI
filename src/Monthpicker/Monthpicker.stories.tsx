import React from 'react';
import { Story, Meta } from 'storybook';
import { Monthpicker, MonthpickerProps } from './Monthpicker';

export default {
    title: 'Development/Monthpicker',
    component: Monthpicker,
} as Meta;

export const Overview: Story<MonthpickerProps> = (args) => {
    return <Monthpicker {...args} />;
};
