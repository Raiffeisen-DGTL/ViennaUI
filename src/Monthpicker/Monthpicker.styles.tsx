import styled from 'styled-components';
import { datepicker } from 'vienna.ui-theme';
import { StyledInput, Wrapper } from '@/Input/Input.styles';
import { getPresets } from '../Utils/styling';

const presets = getPresets(
    datepicker,
    'datepicker'
)({
    calendar: null,
});

export const Box = styled.div`
    position: relative;
    display: flex;
    flex-direction: column;
`;

export const InputBox = styled.div`
    position: relative;

    & ${StyledInput} {
        caret-color: transparent;
        cursor: pointer;
    }
    & ${Wrapper} {
        cursor: pointer;
    }
`;

export const CalendarBox = styled.div`
    ${presets.calendar};
`;
