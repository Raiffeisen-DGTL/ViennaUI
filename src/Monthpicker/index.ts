export { Monthpicker, type MonthpickerProps } from './Monthpicker';
export type { MonthpickerLocalization } from './localization';
