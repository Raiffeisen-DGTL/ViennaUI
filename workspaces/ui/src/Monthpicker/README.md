# Monthpicker

Компонент, который помогает пользователю заполнить дату без учета конкретного дня - месяц и год.


## Импорт

```
import { Monthpicker } from 'vienna-ui';
```

## Свойства / Props

Prop | Type | Default | Description
--- | --- | --- | ---
value | Date \| undefined | false | Выбранная дата
isCalendarOpen | boolean \| undefined | false | Отображение календаря
onChange | ((event: Event \| InputEvent> \| null, data: { value?: string \| undefined; date?: Date \| undefined; name?: string \| undefined; }) => void) \| undefined | false |
design | "outline" \| "material" \| undefined | false | Дизайн
prefix | ReactNode | false | Значанеие отображаемое перед компонентом
autoCapitalize | string \| undefined | false | Автоматическая установка заглавной буквы
onCopy | ((event: ClipboardEvent) => void) \| undefined | false | Обработчик копирования
onCut | ((event: ClipboardEvent) => void) \| undefined | false | Обработчик вырезки
onFocus | InputEvent> \| undefined | false | Обработчик события при получении фокуса компонентом
onBlur | InputEvent> \| undefined | false | Обработчик события при потере фокуса компонентом
onKeyDown | ((event: KeyboardEvent) => void) \| undefined | false | Обработчик события при нажатии кнопки клавиатуры, когда компонент в фокусе
onKeyPress | ((event: KeyboardEvent) => void) \| undefined | false | Обработчик события при нажатии и удержании кнопки клавиатуры с печатемым символом, когда компонент в фокусе
onKeyUp | ((event: KeyboardEvent) => void) \| undefined | false | Обработчик события при отпускании кнопки клавиатуры, когда компонент в фокусе
onMouseDown | Function \| undefined | false | Обработчик события первого полупериода клика
onPointerDown | Function \| undefined | false | Обработчик события первого полупериода клика
size | "xs" \| "s" \| "m" \| "l" \| "xl" \| "xxl" \| undefined | false | Размеры
ref | ((instance: HTMLInputElement \| null) => void) \| RefObject \| null \| undefined | false | Сcылка на нативный элемент input, доступна после отрисовки
invalid | boolean \| undefined | false |
disabled | boolean \| undefined | false |
active | boolean \| undefined | false | Принудительный ховер
postfix | ReactNode | false | Значение отображаемое после компонента
smartPlaceholder | ReactNode | false |
onDispose | (() => void) \| undefined | false |

## Использование

По умолчанию дизайн `outline` и размер `l`.
> Компонент является контролируемым, то есть чтобы отобразить выбранную дату,
> необходимо получить ее значение через обработчик `onChange` и прокинуть в `value`.

```{() => {
    const [value, setValue] = React.useState();
    const handleChange = React.useCallback((e, data) => setValue(data.date), []);
    return <Monthpicker value={value} onChange={handleChange} />;
}}
```

## Выбранная дата
##### Свойство `value`

Выбранное значение передается в формате `Date`.

```
<Monthpicker value={new Date(2020, 2, 4)} />
```

## Размеры
##### Свойство `size`

Компонент имеет стандартные размеры `xs`, `s`, `l` (по умолчанию) и `xl`.

## Дизайн
##### Свойство `design`

Есть два доступных значения для свойства `design`: `outline` (по умолчанию) и `material`.

```jsx
<Monthpicker design='material' />
<Monthpicker design='outline' />
```

## Состояния
##### Свойство `disabled`, `invalid`

Компонент имеет два состояния `disabled` и `invalid`.

```jsx
<Monthpicker value={value} onChange={handleChange} invalid />
<Monthpicker value={value} onChange={handleChange} disabled />
```

## Управление состоянием открытия календаря из родительского компонента
##### Свойство `isCalendarOpen`

Состоянием отображения календаря можно управлять с помощью свойства `isCalendarOpen`.

```jsx
{() => {
    const [value, setValue] = React.useState();
    const handleChange = React.useCallback((e, data) => setValue(data.value), []);
    return <Monthpicker isCalendarOpen={!value} value={value} onChange={handleChange} />;
}}
```

## Элемент, отображаемый в конце поля ввода
##### Свойство `postfix`

Свойство `postfix` управляет отображением элемента в конце поля ввода.
По умолчанию отображается иконка календаря.

```jsx
<Monthpicker postfix={<Icon />} />
```
