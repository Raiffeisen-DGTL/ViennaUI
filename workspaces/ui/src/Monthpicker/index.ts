export * from './Monthpicker';
export type { MonthpickerLocalization } from './localization';
