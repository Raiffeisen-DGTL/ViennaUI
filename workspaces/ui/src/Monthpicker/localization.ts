import { DSLocalizationProps } from '../Localization';
import { CalendarLocalization } from '../Calendar';

export type MonthpickerLocalization = CalendarLocalization;

export type MonthpickerLocalizationProps = DSLocalizationProps<MonthpickerLocalization>;
