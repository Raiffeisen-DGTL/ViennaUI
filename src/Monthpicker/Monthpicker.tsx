import React, { FC, FormEvent, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useFloating, offset, autoPlacement, shift, FloatingPortal } from '@floating-ui/react';
import { format, Locale } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Calendar as CalendarIcon } from 'vienna.icons';
import { CalendarProps } from 'vienna.icons/dist/Calendar/Calendar';
import { Calendar } from '../Calendar';
import { Input, InputEvent, InputProps } from '../Input';
import { Box, InputBox, CalendarBox } from './Monthpicker.styles';
import { MonthpickerLocalizationProps } from './localization';
import { getStringFromDate } from '../Utils/DateUtils';

export interface MonthpickerProps
    extends Omit<InputProps, 'value' | 'type' | 'onChange' | 'onPaste'>,
        MonthpickerLocalizationProps {
    value?: Date;

    /**
     * Отображение календаря
     */
    isCalendarOpen?: boolean;

    onChange?: (
        event: InputEvent<FormEvent<HTMLInputElement>> | Event | null,
        data: { value?: string; date?: Date; name?: string }
    ) => void;
    /**
     * Локаль календаря
     */
    locale?: Locale;
}

export const Monthpicker: FC<MonthpickerProps> = ({
    isCalendarOpen = false,
    name,
    size,
    value,
    onChange,
    onFocus,
    onBlur,
    onKeyPress,
    localization,
    locale,
    ...attrs
}) => {
    const [isOpen, setOpen] = useState<boolean | undefined>(isCalendarOpen);
    const datepickerEl = useRef<HTMLDivElement>(null);
    interface inputType {
        date: Date;
        value: string;
    }

    const initializedValue = useMemo(() => {
        return value ? { date: value, value: format(value, 'LLLL yyyy', { locale: ru }) } : '';
    }, [value]);

    const [inputValue, setInputValue] = useState<inputType | string | (() => inputType)>(initializedValue);
    const [active, setActive] = useState(false);

    const { refs, floatingStyles } = useFloating({
        middleware: [
            offset({
                mainAxis: 4,
                crossAxis: 0,
                alignmentAxis: 0,
            }),
            shift(),
            autoPlacement({
                alignment: 'start',
                allowedPlacements: ['top-start', 'bottom-start'],
                crossAxis: true,
            }),
        ],
    });

    const handleInputFocus = useCallback(
        (event, data) => {
            if (typeof onFocus === 'function') {
                onFocus(event, { value: data.value, name });
            }
            setOpen(true);
        },
        [name, onFocus]
    );

    const handleClickDateIcon = useCallback(
        (e) => {
            e.stopPropagation();
            setOpen(!isOpen);
        },
        [isOpen]
    );

    const handleClickInput = useCallback(() => {
        if (active) {
            setOpen(true);
        }
    }, [isOpen, active]);

    const handleClickMonth = useCallback(
        (event, { date, value }) => {
            if (date) {
                setInputValue({ date, value });
                if (typeof onChange === 'function') {
                    onChange(event, { value: value, date, name });
                }
                setOpen(false);
                setActive(true);
            }
        },
        [name, isOpen, active, onChange]
    );

    const handleChangeInput = useCallback(
        (event, value) => {
            if (typeof onChange === 'function') {
                onChange(event, value);
            }
        },
        [name, onChange, value]
    );

    const handleKeyPress = useCallback(
        (event) => {
            if (typeof onKeyPress === 'function') {
                onKeyPress(event);
            }
            if (event.key === 'Enter') {
                setOpen(!isOpen);
            }
            if (event.key === 'Escape') {
                setOpen(false);
            }
        },
        [isOpen, onKeyPress]
    );

    const handleKeyDown = useCallback((event) => {
        if (event.keyCode === 8) {
            setInputValue('');
            if (typeof onChange === 'function') {
                onChange(event, { date: undefined, value: '', name });
            }
        }
    }, []);

    const handleClickDocument = (event): void => {
        if (
            event.target instanceof Node &&
            datepickerEl &&
            datepickerEl.current &&
            !datepickerEl.current.contains(event.target)
        ) {
            setOpen(false);
            setActive(false);
        }
    };

    const handleBlur = useCallback(
        (event) => {
            if (typeof onBlur === 'function') {
                onBlur(event, {
                    value: typeof value === 'string' ? value : getStringFromDate(value as Date),
                    name,
                });
            }
            setOpen(false);
            setActive(false);
        },
        [onBlur, value, name, active]
    );

    useEffect(() => {
        if (document.body) {
            document.body.addEventListener('click', handleClickDocument);
        }

        return () => {
            if (document.body) {
                document.body.removeEventListener('click', handleClickDocument);
            }
        };
    });

    const handleMouseDownCalendar = useCallback((e) => {
        e.preventDefault();
    }, []);

    return (
        <Box ref={refs.setReference} id='monthpicker-box' onBlur={handleBlur} onClick={handleClickInput}>
            <InputBox>
                <Input
                    {...(attrs as {})}
                    postfix={
                        <CalendarIcon
                            size={size === 'xxl' ? 'xl' : (size as CalendarProps['size'])}
                            cursor='pointer'
                            onClick={handleClickDateIcon}
                        />
                    }
                    active={active}
                    size={size}
                    name={name}
                    value={
                        inputValue &&
                        `${(inputValue as any).value?.split('')[0]?.toUpperCase()}${(inputValue as any).value?.slice(
                            1
                        )}`
                    }
                    placeholder='Выберите месяц'
                    onChange={handleChangeInput}
                    onFocus={handleInputFocus}
                    onKeyPress={handleKeyPress}
                    onKeyDown={handleKeyDown}
                />
            </InputBox>
            {isOpen && (
                <FloatingPortal>
                    <CalendarBox
                        data-id='calendar'
                        ref={refs.setFloating}
                        style={floatingStyles}
                        onMouseDown={handleMouseDownCalendar}>
                        <Calendar
                            mode='month'
                            date={(inputValue as any)?.date}
                            locale={locale}
                            localization={localization}
                            onChangeMonth={handleClickMonth}
                        />
                    </CalendarBox>
                </FloatingPortal>
            )}
        </Box>
    );
};

Monthpicker.displayName = 'Monthpicker';
