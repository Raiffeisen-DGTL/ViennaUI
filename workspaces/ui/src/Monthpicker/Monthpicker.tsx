import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { useDrop } from 'vienna.react-use';
import { format, Locale } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Calendar as CalendarIcon } from 'vienna.icons';
import { Calendar } from '../Calendar';
import { Input, InputEvent, InputProps } from '../Input';
import { Box, InputBox, CalendarBox } from './Monthpicker.styles';
import { MonthpickerLocalizationProps } from './localization';

export interface MonthpickerProps
    extends Omit<InputProps, 'value' | 'type' | 'onChange' | 'onPaste'>,
        MonthpickerLocalizationProps {
    value?: Date;

    /**
     * Отображение календаря
     */
    isCalendarOpen?: boolean;

    onChange?: (
        event: InputEvent<React.FormEvent<HTMLInputElement>> | Event | null,
        data: { value?: string; date?: Date; name?: string }
    ) => void;
    /**
     * Локаль календаря
     */
    locale?: Locale;
}

export const Monthpicker: React.FC<MonthpickerProps> = ({
    isCalendarOpen,
    name,
    size,
    value,
    onChange,
    onFocus,
    onKeyPress,
    localization,
    locale,
    ...attrs
}: MonthpickerProps): JSX.Element => {
    const [isOpen, setOpen] = useState<boolean | undefined>(isCalendarOpen);
    const datepickerEl = useRef<HTMLDivElement>(null);
    const calendarRef = useDrop({ align: 'vertical', float: 'start', margins: { x: 0, y: 4 } });

    const initializedValue = useMemo(() => {
        return value ? { date: value, value: format(value, 'LLLL yyyy', { locale: ru }) } : '';
    }, [value]);
    const [inputValue, setInputValue] = useState<any>(initializedValue);

    const handleInputFocus = useCallback(
        (event, data) => {
            if (typeof onFocus === 'function') {
                onFocus(event, { value: data.value, name });
            }
            setOpen(true);
        },
        [name, onFocus]
    );

    const handleClickDateIcon = () => {
        setOpen(!isOpen);
    };

    const handleClickMonth = useCallback(
        (event, { date, value }) => {
            if (date) {
                setInputValue({ date, value });
                if (typeof onChange === 'function') {
                    onChange(event, { value: value, date, name });
                }
                setOpen(false);
            }
        },
        [name, onChange]
    );

    const handleKeyPress = useCallback(
        (event) => {
            if (typeof onKeyPress === 'function') {
                onKeyPress(event);
            }
            if (event.key === 'Enter') {
                setOpen(!isOpen);
            }
        },
        [isOpen, onKeyPress]
    );

    const handleClickDocument = (event): void => {
        if (
            event.target instanceof Node &&
            datepickerEl &&
            datepickerEl.current &&
            !datepickerEl.current.contains(event.target)
        ) {
            setOpen(false);
        }
    };

    useEffect(() => {
        if (document.body) {
            document.body.addEventListener('click', handleClickDocument);
        }

        return () => {
            if (document.body) {
                document.body.removeEventListener('click', handleClickDocument);
            }
        };
    });

    return (
        <Box ref={datepickerEl}>
            <InputBox>
                <Input
                    postfix={<CalendarIcon size={size === 'xxl' ? 'xl' : size} onClick={handleClickDateIcon} />}
                    size={size}
                    name={name}
                    value={
                        inputValue && `${inputValue.value?.split('')[0]?.toUpperCase()}${inputValue.value?.slice(1)}`
                    }
                    placeholder='Выберите месяц'
                    onFocus={handleInputFocus}
                    onKeyPress={handleKeyPress}
                    {...attrs}
                />
            </InputBox>
            {isOpen && (
                <CalendarBox ref={calendarRef}>
                    <Calendar
                        mode='month'
                        date={inputValue?.date}
                        locale={locale}
                        localization={localization}
                        onChangeMonth={handleClickMonth}
                    />
                </CalendarBox>
            )}
        </Box>
    );
};

Monthpicker.displayName = 'Monthpicker';
Monthpicker.defaultProps = {
    isCalendarOpen: false,
};

export default Monthpicker;
