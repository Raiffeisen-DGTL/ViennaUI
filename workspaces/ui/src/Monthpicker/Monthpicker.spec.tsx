/* eslint-disable @typescript-eslint/no-misused-promises */
import React from 'react';
import { Monthpicker } from './Monthpicker';

test('Monthpicker', async () => {
    const snap = snapshot.render(<Monthpicker />);
    expect(snap).toMatchSnapshot();
});
