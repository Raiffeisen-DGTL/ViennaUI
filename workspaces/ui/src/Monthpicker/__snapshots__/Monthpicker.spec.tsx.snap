// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monthpicker 1`] = `
<div
  class="Monthpickerstyles__Box-sc-2myayb-0 jMyYmn"
>
  <div
    class="Monthpickerstyles__InputBox-sc-2myayb-1 cfspmU"
  >
    <label
      class="Inputstyles__Wrapper-hc8ttv-6 daAZxl"
    >
      <div
        class="Inputstyles__Wrapped-hc8ttv-5 lhoBGY"
      >
        <div
          class="Inputstyles__Box-hc8ttv-1 bzkyqR"
        >
          <input
            aria-invalid="false"
            class="Inputstyles__StyledInput-hc8ttv-2 bUBKOu"
            placeholder="Выберите месяц"
            value=""
          />
        </div>
      </div>
      <div
        class="Inputstyles__Postfix-hc8ttv-4 dvGfwK"
      >
        <svg
          fill="currentColor"
          height="20"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M10 3v2h4V3h2v2h4a1 1 0 011 1v14a1 1 0 01-1 1H4a1 1 0 01-1-1V6a1 1 0 011-1h4V3h2zm8.999 10h-14L5 19h14l-.001-6zM16 15a1 1 0 110 2 1 1 0 010-2zM8 7H5l-.001 4h14L19 7h-3v2h-2V7h-4v2H8V7z"
          />
        </svg>
      </div>
    </label>
  </div>
</div>
`;
